### PR TITLE
ENH: Added passing fa to Projection mappers using Node.pass_attr

### PR DESCRIPTION
**Pull request by swaroopgj**
This mapper will allow adding feature attributes that will be propagated to datasets that are mapped using this mapper. It is useful to keep track of what the new features are after mapping.

### CODE REVIEW

**Comment by swaroopgj** [comment]
Any inputs on whether this feature (of adding fa in mapper) be part of current StaticProjectionMapper or a different mapper as is implemented currently?

**Comment by hanke** [comment]
Sorry for the delay. It seems that the proposed functionality is very similar to what is done by the `pass_attr` argument that any `Node` supports. Do you need extra functionality that is not possible to implement with that facility?

**Comment by swaroopgj** [comment]
Very good point. I think `pass_attr` might suffice. All I need to do is to enable a `ca` in `StaticProjectionMapper` that can be enabled at instantiation and assigned later. Is that true?

**Comment by swaroopgj** [comment]
I experimented with adding `pass_fa` as `ca` in `ProjectionMapper`, but the attributes are only passed when the object is called, but not when using forward(). This makes sense, as the pass_attr is added in `Node._postcall()`. Is there a workaround to this, or do I need to manually call `Node._pass_attr` in `ProjectionMapper._forward_data()`?

**Comment by hanke** [comment]
Hmm, not sure how to do this best.__call__() will execute __forward__() but
not vice versa. Any idea @yarikoptic ?

**Comment by yarikoptic** [comment]
I do not really see a clean way out (yet at least). I guess as an ugly workaround we could overload _postcall in Mappers and move invocation of _pass_attr from there into some *forward

**Comment by swaroopgj** [comment]
Will it be okay to call _pass_attr()  in forward() without changing anything else (for now)?

**Comment by yarikoptic** [comment]
Stop -- why can't you just call the mapper, instead of an explicit .forward()?

**Comment by swaroopgj** [comment]
Good point. Yes, we can just call the object instead. I use `forward` just because it seems intuitive in the code, and other users <i> might also do that </i>. Just to avoid confusion, we should make it explicit in the docs that `pass_attr` only works when calling the object and not with `forward` and people should use call and not forward, preferably. Does that sound good? If so, I can push my changes reflecting that.

**Comment by yarikoptic** [comment]
On Fri, 26 Jun 2015, Swaroop Guntupalli wrote:

>    Good point. Yes, we can just call the object instead. I use forward just
>    because it seems intuitive in the code, and other users might also do that
>    . Just to avoid confusion, we should make it explicit in the docs that
>    pass_attr only works when calling the object and not with forward and
>    people should use call and not forward, preferably. Does that sound good?
>    If so, I can push my changes reflecting that.

sounds good to me!  please just verify that using .call() doesn't
introduce any side-effects ;)

File: mvpa2/base/node.py [hunk 1]
```diff
@@ -84,6 +84,8 @@ def __init__(self, space=None, pass_attr=None, postproc=None, **kwargs):
           'pvalues', while swapping the first and second axes. Simplified
           instructions can be given by leaving out consecutive tuple elements
           starting from the end.
+          Note: For Projection-derived mappers, this only passes attributes when
```
Comment (yarikoptic) [inline mvpa2/base/node.py]: Well -- this would apply to any Mapper, not just Projection-derived, right?

File: mvpa2/mappers/projection.py [hunk 2]
```diff
@@ -69,7 +71,8 @@ def __init__(self, demean=True, **kwargs):
         """Offset (most often just mean) in the input space"""
         self._offset_out = None
         """Offset (most often just mean) in the output space"""
-
+        self.ca.pass_fa = None
+        """Feature attribute to be passed onto output dataset"""
```
Comment (yarikoptic) [inline mvpa2/mappers/projection.py]: oh hoh... hold on... now that I actually read your test which shows the use case.  This is an abuse of .ca which are intended for storage of computational artifacts, thus not to be assigned outside.

What you really want is just to assign that fa in your postproc which you could give to any node. e.g. you could say

```python
def myawkwardcustomfaassignment(ds):
  ds.fa['magicalone'] = ['whatever']*ds.nfeatures
  return ds
```
and then give  to your e.g. SVDMapper   postproc=myawkwardcustomfaassignment

or am I missing the point?

Comment (yarikoptic) [inline mvpa2/mappers/projection.py]: ie, you don't need pass_attr at all... or am I missing the point?

**Comment by swaroopgj** [comment]
You are correct. I can use postproc for my use case. Closing this PR.

**Pull request closed without merging**